Port to Greenplum

liuxueyang opened the pull request:
- Fix Makefile and unittest cases for gpdb
- Use the random distribution policy when creating tables
- Use CdbDispatch in hll_set_max_sparse UDF. In this way, the user
doesn't need to use `gp_dist_random` when calling this UDF
- Remove upgrade extension statements in the `setup.sql` test case

Pick changes from https://github.com/greenplum-db/gpdb/pull/13658

Co-authored-by: Xuebin Su <sxuebin@vmware.com>
Co-authored-by: Sasasu <i@sasa.su>

liuxueyang committed:
Port to Greenplum

- Fix Makefile and unittest cases for gpdb
- Use random distribution policy when creating table
- Use CdbDispatch in hll_set_max_sparse UDF. In this way, the user
doesn't need to use `gp_dist_random` when call this UDF
- Remove upgrade extension statements in the `setup.sql` test case

Sasasu commented on src/hll.c:
```diff
@@ -2824,12 +2826,21 @@ hll_set_max_sparse(PG_FUNCTION_ARGS)
                 (errcode(ERRCODE_DATA_EXCEPTION),
                  errmsg("sparse threshold must be in range [-1,MAXINT]")));
 
+    if (IS_QUERY_DISPATCHER())
```
I remember @kainwen said `IS_QUERY_DISPATCHER` does not work well when `SingleQE`.

but I think `if (GpIdentity.segindex == MASTER_CONTENT_ID)` is really what we want (modify this GUC on all host when the query is only executed on coordinator).

does he check this?

kainwen replied:
We need to check `Gp_role` is `GP_ROLE_DISPATCH`, because only QD can do dispatch.

kainwen commented on expected/add_agg.out:
```diff
@@ -11,7 +11,7 @@ DROP TABLE IF EXISTS test_khvengxf;
 CREATE TABLE test_khvengxf (
 	val    integer
-);
+) DISTRIBUTED RANDOMLY;
```
Is there any specific reason that we have to force randomly here?

liuxueyang replied:
No, there isn't any specific reason. I just set the random distribution policy as default when creating tables in this commit: https://github.com/greenplum-db/postgresql-hll/pull/2/commits/38838325f0ae0a2231f8a971984562462f0679cf. Maybe this will cover some MPP cases?

kainwen commented on update/hll--2.10.sql:
```diff
@@ -271,7 +271,7 @@ CREATE FUNCTION hll_set_output_version(integer)
      RETURNS integer
      AS 'MODULE_PATHNAME'
-     LANGUAGE C STRICT IMMUTABLE;
+  LANGUAGE C STRICT;
```
Why we remove `immutable`?

liuxueyang replied:
If it's immutable, then the optimizer will optimize it to constant in this statement: `select hll_set_max_sparse(0) from gp_dist_random('gp_id');`. But I modify the C code of this UDF later. So this modification is not needed anymore. I'll restore it.

xuebinsu replied:
Is the function still consider IMMUTABLE if the behavior is different when it is executed on QD from when it is executed on QE?

liuxueyang replied:
According to the `immutable` classification: `An IMMUTABLE function cannot modify the database and is guaranteed to return the same results given the same arguments forever.` It doesn't mention the different behavior of the QD and QE. I'm not sure either.

xuebinsu replied:
True. Since its result depends on where it is executed and is not guaranteed to be the same "forever", I tend to believe it is not IMMUTABLE.

liuxueyang committed:
Setting GUC gp_create_table_random_default_distribution

Set this variable on database level. So we don't need to add
`DISTRIBUTED RANDOMLY` when creating tables.

liuxueyang committed:
Add immutable back to hll_set_max_sparse

I modify the C implementation of this UDF, so it's ok to be immutable

liuxueyang committed:
Add init_file to ignore some patterns

This will remove those 'HINT:' changes in the `.out` file

yihong0618 commented on sql/setup.sql:
```diff
@@ -1,7 +1,12 @@
+CREATE EXTENSION hll VERSION '2.16';
+
+-- Using random distribution policy by default when creating tables
+ALTER database contrib_regression
+  SET gp_create_table_random_default_distribution TO 'on';
```
do we need to set it back to default in the end?

liuxueyang replied:
No. At the beginning of test, the `contrib_regression` database is dropped and recreated again.

kainwen replied:
> No. At the beginning of test, the `contrib_regression` database is dropped and recreated again.

Will it impact other tests under contrib?

And I still do not get the point to set this GUC. Do you find we set it for other test cases? I think this GUC is not frequently used.

liuxueyang replied:
Using the random distribution policy, in this case, may not make much sense. Maybe using the default distribution policy is enough?

kainwen replied:
> Using the random distribution policy, in this case, may not make much sense. Maybe using the default distribution policy is enough?

Normally we do not change this. If nothing special, we can just let it act as default behavior.

liuxueyang replied:
I use this UDF to avoid specifying `distributed randomly` in every `create table`.

yihong0618 commented on sql/setup.sql:
```diff
@@ -1,7 +1,12 @@
+ALTER database contrib_regression
+  SET gp_create_table_random_default_distribution TO 'on';
+
+-- ALTER EXTENSION hll UPDATE TO '2.11';
+-- ALTER EXTENSION hll UPDATE TO '2.12';
+-- ALTER EXTENSION hll UPDATE TO '2.13';
+-- ALTER EXTENSION hll UPDATE TO '2.14';
+-- ALTER EXTENSION hll UPDATE TO '2.15';
+-- ALTER EXTENSION hll UPDATE TO '2.16';
```
maybe we can add some comments why we `comments this lines`?

liuxueyang replied:
Comments added.

yihong0618 commented on sql/disable_hashagg.sql:
```diff
@@ -180,6 +180,5 @@ FROM
 	add_test_table 
 GROUP BY 1;
 
-RESET work_mem;
```
we do not need to reset?

liuxueyang replied:
Each SQL script is run in a separate session and `statement_mem` is a session-level parameter. But it's ok to `RESET` it too.

yihong0618 commented on expected/disable_hashagg_1.out:
```diff
@@ -29,6 +29,8 @@ SELECT hll_set_output_version(1);
 (1 row)
 
 CREATE TABLE tt1(id int, col_1 int, sd hll);
+NOTICE:  using default RANDOM distribution since no distribution was specified
+HINT:  Consider including the 'DISTRIBUTED BY' clause to determine the distribution of rows.
```
this can delete too?I think

liuxueyang replied:
There are some other diffs in this output file. I think removing these two lines by hand is not appropriate.

liuxueyang committed:
Use Gp_role to check QD

liuxueyang committed:
Add comment to sql/set.sql

kainwen commented on src/hll.c:
```diff
@@ -2824,12 +2826,21 @@ hll_set_max_sparse(PG_FUNCTION_ARGS)
+    {
+        initStringInfo(&sql);
+        appendStringInfo(&sql, "select hll_set_max_sparse(%d);", maxsparse);
+        CdbDispatchCommand(sql.data, DF_NONE, NULL);
```
`DF_NONE` means no distributed snapshot, this is intended?

liuxueyang replied:
In this case, the UDF sets a global variable conf. I guess a distributed snapshot is not needed.

liuxueyang committed:
Use default distribution policy

liuxueyang committed:
Update testcase outputs for default distribution policy

liuxueyang committed:
RESET statement_mem

yihong0618 reviewed (approved):
LGTM
